okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -257,6 +257,29 @@ private RealConnection findConnection(int connectTimeout, int readTimeout, int w
     return result;
   }
 
+  /**
+   * Returns the previous allocated connection or null if there is no previous allocated connection.
+   * We will have a previous allocated connection in the case of follow-up requests. With HTTP/2
+   * multiple requests share the same connection so it's possible that our connection is flagged to
+   * disallow new streams during a follow-up request.
+   */
+  private RealConnection previousAllocatedConnection() {
+    assert (Thread.holdsLock(connectionPool));
+
+    RealConnection allocatedConnection = this.connection;
+    if (allocatedConnection == null) return null;
+
+    if (allocatedConnection.noNewStreams) {
+      // Our connection was flagged to disallow new streams. This could happen in HTTP/2 when
+      // multiple requests use the same connection. Discard it.
+      deallocate(false, false, true);
```

Comment:
Maybe instead of calling this method `previousAllocatedConnection()` this method should instead be called say, `releaseIfNoNewStreams()` and that could return a `Socket` ?